transport/cluster.go

```diff
@@ -0,0 +1,295 @@
+				Addr:       addr,
+				Datacenter: string(v[nodeDC]),
+				Rack:       string(v[nodeRack]),
+				Tokens:     v[nodeTokens],
+			}
+			if pool, err := NewConnPool(appendDefaultPort(addr), c.cfg); err != nil {
+				n.Status = atomic.NewBool(statusDown)
+			} else {
+				n.Status = atomic.NewBool(statusUP)
+				n.Pool = pool
+			}
+			m[addr] = n
+		}
+	}
+
+	for k, v := range old {
+		if _, ok := m[k]; !ok {
+			v.Pool.Close()
+		}
+	}
+	c.SetPeers(m)
+	return nil
+}
+
+var (
+	peerQuery = Statement{
+		Content:     "SELECT peer, data_center, rack, tokens FROM system.peers",
+		Consistency: frame.ONE,
+	}
+
+	localQuery = Statement{
+		Content:     "SELECT rpc_address, data_center, rack, tokens FROM system.local",
+		Consistency: frame.ONE,
+	}
+)
+
+// Indexes of columns returned from refresh topology queries.
+const (
+	nodeAddr   = 0
+	nodeDC     = 1
+	nodeRack   = 2
+	nodeTokens = 3
+)
+
+func (c *Cluster) getAllNodesInfo() ([]frame.Row, error) {
+	peerRes, err := c.control.Query(peerQuery, nil)
+	if err != nil {
+		return nil, fmt.Errorf("discovering peer topology: %w", err)
+	}
+
+	localRes, err := c.control.Query(localQuery, nil)
+	if err != nil {
+		return nil, fmt.Errorf("discovering local topology: %w", err)
+	}
+
+	return append(peerRes.Rows, localRes.Rows[0]), nil
+}
+
+func (c *Cluster) GetPeers() PeerMap {
+	return c.peers.Load().(PeerMap)
+}
+
+func (c *Cluster) SetPeers(m PeerMap) {
+	c.peers.Store(m)
+}
+
+const defaultPort = 9042
+
+func appendDefaultPort(addr string) string {
+	return addr + ":" + strconv.Itoa(defaultPort)
+}
+
+const eventChanSize = 32
+
+func (c *Cluster) handleEvents() {
+	for {
+		res := <-c.events
+		if res.Err != nil {
+			return
+		}
+		switch v := res.Response.(type) {
+		case *TopologyChange:
+			c.handleTopologyChange(v)
+		case *StatusChange:
+			c.handleStatusChange(v)
+		case *SchemaChange:
+			// TODO: add schema change.
+		default:
+			log.Printf("unsupported event type: %v", res.Response)
+		}
+	}
+}
+
+func (c *Cluster) handleTopologyChange(v *TopologyChange) {
+	log.Printf("handling node: %s topology change to: %s", v.Address.String(), v.Change)
+	c.RequestRefresh()
+}
+
+func (c *Cluster) handleStatusChange(v *StatusChange) {
+	log.Printf("handling node: %s status change to: %s", v.Address.String(), v.Status)
+	m := c.GetPeers()
+	addr := v.Address.String()
+	if n, ok := m[addr]; ok {
+		switch v.Status {
+		case frame.Up:
+			n.SetStatus(statusUP)
+		case frame.Down:
+			n.SetStatus(statusDown)
+		default:
+			log.Fatalf("status change not supported: %s", v.Status)
+		}
+	} else {
+		log.Printf("node which status is being set to %s was not present in known topology: %s", v.Status, addr)
+		c.RequestRefresh()
+	}
+}
+
+// RequestRefresh notifies cluster that it should update topology.
+// TODO: do we need some mechanism for notifying requester that refresh was successful?
+func (c *Cluster) RequestRefresh() {
+	c.refresher <- struct{}{}
+}
+
+func (c *Cluster) StopEventHandling() {
```

Comment:
Should't this be called Close?

Reply:
Is there StartEventHandling

Reply:
@Michal-Leszczynski there is a fundamental BUG(s) here!!

- handleEvents is not prepared for this to close
- you are closing something you do not own
- conn will panic when you close it

Reply:
That's true and I actually spotted this bug because of TestClusterIntegration!
I already fixed it, but forgot to delete this useless function.

Reply:
See the final statement https://github.com/mmatczuk/scylla-go-driver/pull/130#pullrequestreview-907045101